Add more support for (v,e) TabulatedFP

GiudGiud opened the pull request:
More oob options
Easier tuning of domain of validity
Loading direct (v,e) as PC can generate those tabulations directly
An option for not needing (v,h) interpolations, only relying on (v,e) and (p,T)

Refs #20101

GiudGiud committed:
Allow user to specify emin/max and vmin/vmax bounds directly
This will simplify the logic which is obtained from dubious conversions from (p,T) at the very extreme of the domain (p=pmin, T=Tmin) for example
refs #20101

GiudGiud committed:
Add input variable checks on (v,e) variable set
refs #20101

GiudGiud committed:
Add more options for out-of-bounds behavior treatments
refs #20101

GiudGiud committed:
Regold for better behavior on g(v,e) call (using v,e bounds rather than invalid OOB values)
Use closest bound when setting TFP solution as invalid (to be able to keep going and not crash the newton solves)
refs #20101

GiudGiud committed:
Re-order attributes for clarity:
- group attributes that play symmetric roles in (p,T) and (v,e)
- add symmetric parameters for (v,e) tabulations I/O
refs #20101

GiudGiud committed:
Add a direct interpolation path to all _from_v_e tabulated calls

GiudGiud committed:
Use the booleans to keep track of whether to interpolate. NOTE that we
do not forward (v,e) calls to _fp directly, instead we prefer to compute (p,T) from (v,e) then forward to the _fp (p,T) calls
This is because we expect several FPs to not have (v,e) calls implemented

GiudGiud committed:
Symmetrize treatment of pT and ve interpolations
This is done because a user may want only (v,e) interpolations, and should not get errors from missing data for pT

GiudGiud committed:
Add ve tabulation source data generation
Clean up error logic as having a _fp object is guaranteed

GiudGiud committed:
Move (v,e) grid generation code to base TFP class to be able to use it for generating (v,e) data from
a fluid property object

GiudGiud committed:
Allow for reading (v,e) tabulations directly
- duplicate code specific to reading pT
- commonalize code shared between reading pT and ve

GiudGiud committed:
Avoid generating pT interpolations without data
Add s(v,e) routine to avoid using pT/ve inversions when using v,e direct interpolations
Add a _fp path for all (v,e) calls when not interpolating those (same as for (p,T) calls)
Add individual errors in (v,e) calls when nothing can compute the value (same as for (v,e) calls)
Add support for reading the columns needed for (v,e) interpolation
More rigorous input selections checking

GiudGiud committed:
Add test for direct from (v,e) interpolations

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2320494/) on 0f6b54f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28085/site)

This comment will be updated on new commits.

GiudGiud committed:
Address test failures:
- initialize p,T to avoid compiler warning (undeserved, the function call would have set them)
- regold due to adding s_from_v_e to Tabulated instead of conversions through spT
- fix mistake in indexing for unequal column error
- dont use recover testing on steady solves
- use different suffix to prevent test results interference

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2321365/) on e1ae0ab wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f75640</th>
      <th colspan=4>#28085 e1ae0a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.06%</td>
      <td>85.06%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104534</td>
      <td>104533</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18360</td>
      <td>18361</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28085/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28085/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>f75640</th>
      <th colspan=4>#28085 e1ae0a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.61%</td>
      <td>84.48%</td>
      <td>-1.14%</td>
      <td>70.63%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7337</td>
      <td>7504</td>
      <td>+167</td>
      <td>416</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1233</td>
      <td>1379</td>
      <td>+146</td>
      <td>173</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28085/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28085/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28085/coverage/xfem)

</details>

## Warnings

- `fluid_properties` new line coverage rate 70.63% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
Failure unrelated

GiudGiud committed:
Add a Newton solve from ve data to be able to compute e_from_v_h

GiudGiud committed:
Add a h_from_v_e for simplicity
Add output of the interpolation so we can rely on fp for
h_from_v_e, which is used in e_from_v_h

GiudGiud committed:
Add h_from_v_e for simplicity

GiudGiud committed:
Add unit tests for new direct from v,e path

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2326591/) on ebda35b wanted to post the following:

The following commit(s) contain fixups:

- ebda35bdf2: fixup! Add unit tests for new direct from v,e path
- b8c4b91679: fixup! Add h_from_v_e for simplicity

GiudGiud commented:
Adding some basic (v,h) support based on feedback. Previous to this is ready for review

GiudGiud committed:
Rename the FP classes attributes in TFP unit test for clarity
Use IdealGas as reference to be able to test more properties and get less conversion failures

GiudGiud committed:
Use booleans to prevent seg faulting into interpolation array
Fix math for h = e + pv
Use finite differencing to get e_from_v_h derivative, for now

GiudGiud committed:
Add documentation for direct (v,e) interpolations

GiudGiud committed:
Enable reading different files in the CSV reader after the 1st one

lindsayad commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -1411,8 +1717,397 @@ TabulatedFluidProperties::checkInitialGuess() const
+    // if csv exists, get max and min values from csv file
+    else
+    {
+      _e_min = *min_element(_properties[_internal_energy_idx].begin(),
```
```suggestion
      _e_min = *std::min_element(_properties[_internal_energy_idx].begin(),
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -1411,8 +1717,397 @@ TabulatedFluidProperties::checkInitialGuess() const
+    {
+      _e_min = *min_element(_properties[_internal_energy_idx].begin(),
+                            _properties[_internal_energy_idx].end());
+      _e_max = *max_element(_properties[_internal_energy_idx].begin(),
```
```suggestion
      _e_max = *std::max_element(_properties[_internal_energy_idx].begin(),
```

lindsayad commented on modules/fluid_properties/test/tests/tabulated/tabulated.i:
did anything actually change in this file? Can't wait until we've done a PR with formatting, so that we don't see these diffs anymore

lindsayad commented on modules/fluid_properties/test/tests/tabulated/tests:
```diff
@@ -81,19 +83,36 @@
+                FluidProperties/tabulated/fluid_property_ve_file='gold/fluid_properties_out_ve.csv'
+                Outputs/file_base=tabulated_direct_v_e_out"
+    rel_err = 1e-4
+    threading = '!pthreads'
```
what's the reason for the restriction?

GiudGiud replied:
this predates me, I m just copy pasting

GiudGiud committed:
Address Alex review

use libmesh routine for static cast
use std namespace explicitly

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -936,170 +946,371 @@ TabulatedFluidProperties::vaporPressure(Real temperature, Real & psat, Real & dp
+  else if (_fp)
+    return _fp->p_from_v_e(v, e);
+  else
+    mooseError(__PRETTY_FUNCTION__,
```
traces are not included in opt mode iirc
I like calling out the routine for an unimplemented error

GiudGiud commented on modules/fluid_properties/test/tests/tabulated/tabulated.i:
```diff
@@ -11,111 +11,109 @@
     fp = co2
     interpolated_properties = 'density enthalpy viscosity internal_energy k c cv cp entropy'
     # fluid_property_file = fluid_properties.csv
-    save_file = true
     construct_pT_from_ve = false
     construct_pT_from_vh = false
-    # error_on_out_of_bounds = false
```
this changed in that file.
both the parameters are deprecated

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2344319/) on 53f0abf wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28085/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28085/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4d623c8daa651cad6ba44e7772083f5a92be1172`

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -884,26 +850,70 @@ TabulatedFluidProperties::s_from_p_T(Real p, Real T, Real & s, Real & ds_dp, Rea
+    // Lambda computes h from v and the current_e
+    auto lambda = [&](Real v, Real current_e, Real & new_h, Real & dh_dv, Real & dh_de)
+    { h_from_v_e(v, current_e, new_h, dh_dv, dh_de); };
+    const auto e_data =
```
you're not wrong
I ll do this all over FPs next time I touch them if you don't mind

GiudGiud commented:
Thanks for the review!
